ArmPkg/Drivers/CpuDxe/PageTableMemoryAllocation.c

```diff
@@ -0,0 +1,193 @@
+/** @file PageTableAlloc.c
+
+Logic facilitating the pre-allocation of page table memory.
+
+Copyright (C) Microsoft Corporation. All rights reserved.
+SPDX-License-Identifier: BSD-2-Clause-Patent
+**/
+
+#include <Uefi.h>
+#include <Library/DebugLib.h>
+#include <Library/MemoryAllocationLib.h>
+#include <Library/UefiBootServicesTableLib.h>
+#include <Protocol/ArmPageTableMemoryAllocation.h>
+
+#pragma pack(1)
+
+typedef struct {
+  VOID     *NextPool;
+  UINTN    Offset;
+  UINTN    FreePages;
+} PAGE_TABLE_POOL;
+
+#pragma pack()
+
+#define MIN_PAGES_AVAILABLE  5
+
+UINTN            TotalFreePages     = 0;
+PAGE_TABLE_POOL  *mPageTablePool    = NULL;
+BOOLEAN          mPageTablePoolLock = FALSE;
+
+/**
+  Allocates pages for page table memory and adds them to the pool list.
+
+  @param[in] PoolPages The number of pages to allocate. The actual number of pages
+                       reserved will be at least 512 - the number of pages required to describe a
+                       fully split 2MB region.
+
+  @retval EFI_SUCCESS             More pages successfully added to the pool list
+  @retval EFI_INVALID_PARAMETER   PoolPages is zero
+  @retval EFI_ABORTED             Software lock is held by another call to this function
+  @retval EFI_OUT_OF_RESOURCES    Unable to allocate pages
+**/
+EFI_STATUS
+GetMorePages (
+  IN  UINTN  PoolPages
+  )
+{
+  VOID        *Buffer;
+  EFI_STATUS  Status = EFI_SUCCESS;
+
+  if (PoolPages == 0) {
+    return EFI_INVALID_PARAMETER;
+  }
+
+  PoolPages++;
+  PoolPages = ALIGN_VALUE (PoolPages, EFI_SIZE_TO_PAGES (SIZE_2MB)); // Add one page for the header
+  Buffer    = AllocateAlignedPages (PoolPages, BASE_2MB);
+  if (Buffer == NULL) {
+    DEBUG ((DEBUG_ERROR, "ERROR: Out of aligned pages\r\n"));
+    Status = EFI_OUT_OF_RESOURCES;
+    goto Done;
+  }
+
+  if (mPageTablePool == NULL) {
+    mPageTablePool           = Buffer;
+    mPageTablePool->NextPool = mPageTablePool;
+  } else {
+    ((PAGE_TABLE_POOL *)Buffer)->NextPool = mPageTablePool->NextPool;
+    mPageTablePool->NextPool              = Buffer;
+    mPageTablePool                        = Buffer;
+  }
+
+  //
+  // Reserve one page for pool header.
+  //
+  mPageTablePool->FreePages = PoolPages - 1;
+  mPageTablePool->Offset    = EFI_PAGE_SIZE;
+  TotalFreePages           += mPageTablePool->FreePages;
+
+Done:
+  return Status;
+}
+
+/**
+  Allocate memory for the page table.
+
+  @param[in]  NumPages          Number of pages to allocate.
+
+  @retval A pointer to the allocated page table memory or NULL if the
+          allocation failed.
+**/
+VOID *
+AllocatePageTableMemory (

```

Comment:
How would this mechanism prevent re-entrant calls if memory manager need to apply page table attributes in place and free page table buffer ran out?

Reply:
If we're going to go below the minimum number of pages reserved by serving the current request, the logic will hold a semaphore and call to allocate more page table memory. That request will require at most 4 more pages if every level needs a table entry and may cause reentry into AllocatePageTableMemory(). Because the semaphore is held, that request will be serviced by the pages still in reserve to prevent infinite recursion. Once the top level request for more translation table memory resumes execution, it will relinquish the lock and have at least another 512 pages for translation table entries.